### PR TITLE
msg/async/rdma: fix potential chunk recall leak

### PR DESCRIPTION
**Pull request by yuyuyu101**
Signed-off-by: Haomai Wang <haomai@xsky.com>

### CODE REVIEW

File: src/msg/async/rdma/RDMAConnectedSocketImpl.cc [hunk 1]
```diff
@@ -208,10 +208,11 @@ ssize_t RDMAConnectedSocketImpl::read(char* buf, size_t len)
     if (response->byte_len == 0) {
       if (connected) {
         error = ECONNRESET;
-        assert(infiniband->post_chunk(chunk) == 0);
+        cleanup();
         ldout(cct, 20) << __func__ << " got remote close msg..." << dendl;
       }
-      break;
+      assert(infiniband->post_chunk(chunk) == 0);
```
Comment (tchaikov) [inline src/msg/async/rdma/RDMAConnectedSocketImpl.cc]: please don't rely on the side effect of `assert()`.

Comment (yuyuyu101) [inline src/msg/async/rdma/RDMAConnectedSocketImpl.cc]: suggestion?

Comment (tchaikov) [inline src/msg/async/rdma/RDMAConnectedSocketImpl.cc]:
```c++
  int r = infiniband->post_chunk(chunk));
  assert(!r);
```
or simply `ceph_assert_always()` maybe?

**Comment by Adirl** [comment]
Looks good to me.

**Comment by yuyuyu101** [comment]
@tchaikov

File: src/msg/async/rdma/RDMAStack.h [hunk 2]
```diff
@@ -185,9 +185,24 @@ class RDMAConnectedSocketImpl : public ConnectedSocketImpl {
 
  public:
   RDMAConnectedSocketImpl(CephContext *cct, Infiniband* ib, RDMADispatcher* s,
-                          RDMAWorker *w);
+                          RDMAWorker *w)
```
Comment (tchaikov) [inline src/msg/async/rdma/RDMAStack.h]: does this compile?

**Comment by yuyuyu101** [comment]
@tchaikov ....sorry, wrong commit

**Comment by yuyuyu101** [comment]
done

File: src/msg/async/rdma/RDMAConnectedSocketImpl.cc (hunk 1 above)
Comment (tchaikov) [inline src/msg/async/rdma/RDMAConnectedSocketImpl.cc]: comment not addressed.

File: src/msg/async/rdma/RDMAConnectedSocketImpl.cc [hunk 3]
```diff
@@ -518,6 +514,13 @@ void RDMAConnectedSocketImpl::cleanup() {
     delete con_handler;
     con_handler = nullptr;
   }
+  Mutex::Locker l(lock);
+  for (unsigned i=0; i < wc.size(); ++i)
+    assert(infiniband->recall_chunk(reinterpret_cast<Chunk*>(wc[i].wr_id)));
```
Comment (tchaikov) [inline src/msg/async/rdma/RDMAConnectedSocketImpl.cc]: ditto.

File: src/msg/async/rdma/RDMAConnectedSocketImpl.cc [hunk 4]
```diff
@@ -518,6 +514,13 @@ void RDMAConnectedSocketImpl::cleanup() {
...
+  wc.clear();
+  for (auto chunk : buffers)
+    assert(infiniband->post_chunk(chunk) == 0);
```
Comment (tchaikov) [inline src/msg/async/rdma/RDMAConnectedSocketImpl.cc]: here also.

File: src/msg/async/rdma/RDMAConnectedSocketImpl.cc [hunk 5]
```diff
@@ -518,6 +514,13 @@ void RDMAConnectedSocketImpl::cleanup() {
     delete con_handler;
     con_handler = nullptr;
   }
+  Mutex::Locker l(lock);
+  for (unsigned i=0; i < wc.size(); ++i)
```
Comment (tchaikov) [inline src/msg/async/rdma/RDMAConnectedSocketImpl.cc]: add spaces around `=`

**Comment by yuyuyu101** [comment]
updated

**Comment by Adirl** [comment]
so far using this code caused OSD crash
need more testing...

**Pull request closed without merging**